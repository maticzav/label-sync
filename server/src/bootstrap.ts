import { compile } from 'handlebars'

import { mapEntries } from './data/dict'
import { GHTree } from './github'

/**
 * Populates the template with repositories.
 * @param tree
 */
export function populateTempalte(
  tree: GHTree,
  data: { repository: string; repositories: { name: string }[] },
): GHTree {
  return mapEntries(tree, (file) => compile(file)(data))
}
