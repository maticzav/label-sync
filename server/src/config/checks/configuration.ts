import ml from 'multilines'
import * as os from 'os'

import * as t from '../types'
import { Check } from '../parser'

import { mapEntries, mapKeys } from '../../data/map'
import { withDefault } from '../../utils'

export const checkConfiguration: Check = (config) => {
  /**
   * We accumulate possible errors in all the checks and throw the collection.
   */
  let errors: string[] = []

  /* MISSING SIBLINGS */

  /**
   * All siblings should be defined in the configuration.
   */

  let missingSiblings: { repo: string; sibling: string; label: string }[] = []
  for (const repoName in config.repos) {
    const repoConfig = config.repos.get(repoName)!
    const configuredLabels = Object.keys(repoConfig.labels)

    /* Check every label for unconfigured siblings. */
    for (const label in repoConfig.labels) {
      const siblings = withDefault([], repoConfig.labels.get(label)?.siblings)

      for (const sibling of siblings) {
        /* Report unconfigured missing sibling. */
        if (!configuredLabels.includes(sibling)) {
          missingSiblings.push({ repo: repoName, sibling, label })
        }
      }
    }
  }

  if (missingSiblings.length !== 0) {
    const report = ml`
    | Configuration references unconfigured siblings:
    | ${missingSiblings
      .map((s) => `* \`${s.sibling}\` in ${s.repo}:${s.label}`)
      .join(os.EOL)}
    `
    errors.push(report)
  }

  /* ALIAS SPLITING */

  /**
   * You shouldn't make two new labels aliasing one label.
   */

  let duplicateAliaii: { repo: string; alias: string; labels: string[] }[] = []

  /* Check each repository */
  for (const repoName in config.repos) {
    const repoConfig = config.repos.get(repoName)!

    let aliaiiInRepo: { [alias: string]: string[] } = {}

    /* Check every label for unconfigured siblings. */
    for (const label in repoConfig.labels) {
      const aliaii = withDefault([], repoConfig.labels.get(label)?.alias)

      /* Create a map of aliases. */
      for (const alias of aliaii) {
        if (!aliaiiInRepo.hasOwnProperty(alias)) aliaiiInRepo[alias] = []
        aliaiiInRepo[alias].push(label)
      }
    }

    /* Find multiples */
    for (const alias in aliaiiInRepo) {
      if (aliaiiInRepo[alias].length > 1) {
        duplicateAliaii.push({
          repo: repoName,
          alias,
          labels: aliaiiInRepo[alias],
        })
      }
    }
  }

  if (duplicateAliaii.length !== 0) {
    const report = ml`
    | Configuration references same old label twice in alias:
    | ${duplicateAliaii
      .map(
        (s) => `* \`${s.alias}\` appears in ${s.repo}:${s.labels.join(', ')}`,
      )
      .join(os.EOL)}
    |
    | Each legacy label may only be referenced once in alias.
    `
    errors.push(report)
  }

  /* SELF REFERENCES */

  /**
   * Siblings shouldn't self-reference themselves.
   */

  let selfreferencingSiblings: { repo: string; label: string }[] = []
  for (const repoName in config.repos) {
    const repoConfig = config.repos.get(repoName)!

    /**
     * Check every label if it references itself.
     */
    for (const label in repoConfig.labels) {
      const siblings = withDefault([], repoConfig.labels.get(label)?.siblings)

      if (siblings.includes(label)) {
        selfreferencingSiblings.push({ repo: repoName, label })
      }
    }
  }

  if (selfreferencingSiblings.length !== 0) {
    const report = ml`
    | Some of the labels in the configuration reference itself as a sibling:
    | ${selfreferencingSiblings
      .map((s) => `* \`${s.label}\` in ${s.repo}`)
      .join(os.EOL)}
    `
    errors.push(report)
  }

  /* DESCRIPTION LENGTHS */

  /**
   * Siblings shouldn't self-reference themselves.
   */

  let labelsWithTooLongDescriptions: { repo: string; label: string }[] = []
  for (const repoName in config.repos) {
    const repoConfig = config.repos.get(repoName)!

    /**
     * Check every label if it references itself.
     */
    for (const label in repoConfig.labels) {
      const { description } = repoConfig.labels.get(label)!

      if (description && description.length > 1000) {
        labelsWithTooLongDescriptions.push({ repo: repoName, label })
      }
    }
  }

  if (labelsWithTooLongDescriptions.length !== 0) {
    const report = ml`
     | Some of the labels in the configuration have too long descriptions:
     | ${selfreferencingSiblings
       .map((s) => `* \`${s.label}\` in ${s.repo}`)
       .join(os.EOL)}
      
     | The maximum is 1000 characters.
     `
    errors.push(report)
  }

  /**
   * Collect the errors and throw if necessary.
   */

  if (errors.length > 0) {
    const report = errors.join(os.EOL)

    return { success: false, error: report }
  }

  return { success: true, config }
}

// MARK: - Utility functions

/**
 * TRANSFORMATIONS
 */

/**
 * Catches configuration issues that io-ts couldn't catch.
 */
function fixConfig(config: t.LSCConfiguration): t.LSCConfiguration {
  const repos = mapEntries(config.repos, fixRepoConfig)
  const lowercaseRepos = mapKeys(repos, (repo) => repo.toLowerCase())

  return { ...config, repos: lowercaseRepos }
}

/**
 * Fixes issues that io-ts couldn't catch.
 */
function fixRepoConfig(config: t.LSCRepository): t.LSCRepository {
  const labels = mapEntries(config.labels, (label) => ({
    ...label,
    color: fixLabelColor(label.color),
  }))

  return { ...config, labels }
}

/**
 * Removes hash from the color.
 */
export function fixLabelColor(color: string): string {
  if (color.startsWith('#')) return color.slice(1).toLowerCase()
  return color.toLowerCase()
}
