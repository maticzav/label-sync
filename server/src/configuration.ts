import * as t from 'io-ts'
import { PathReporter } from 'io-ts/lib/PathReporter'
import yaml from 'js-yaml'
import * as os from 'os'

import { mapEntries } from './data/dict'

/**
 * Configuration repository is the repository which LabelSync
 * uses to determine the configuration of the service.
 *
 * @param organization
 */
export const getLSConfigRepoName = (owner: string) => `${owner}-labelsync`

/**
 * Configuration file path determines the path of the file in the repositoy
 * that we use to gather configuration information from.
 *
 * It should always be in YAML format.
 */
export const LS_CONFIG_PATH = 'labelsync.yml'

/**
 * Label represents the central unit of LabelSync. Each label
 * can have multiple siblings that are meaningfully related to
 * a label itself, multiple hooks that trigger different actions.
 */
const LSCLabel = t.intersection([
  t.type({
    color: t.string,
  }),
  t.partial({
    description: t.string,
    siblings: t.array(t.string),
    alias: t.array(t.string),
  }),
])
export type LSCLabel = t.TypeOf<typeof LSCLabel>

const LSCLabelName = t.string
export type LSCLabelName = t.TypeOf<typeof LSCLabelName>

/**
 * Repository configuration for how LabelSync should sync it.
 */
const LSCRepositoryConfiguration = t.partial({
  removeUnconfiguredLabels: t.boolean,
})
export interface LSCRepositoryConfiguration
  extends t.TypeOf<typeof LSCRepositoryConfiguration> {}

/**
 * Repository represents a single Github repository.
 * When configured as `strict` it will delete any surplus of labels
 * in the repository.
 */
const LSCRepository = t.intersection([
  t.partial({
    config: LSCRepositoryConfiguration,
  }),
  t.type({
    labels: t.record(LSCLabelName, LSCLabel),
  }),
])
export interface LSCRepository extends t.TypeOf<typeof LSCRepository> {}
const LSCRepositoryName = t.string
export type LSCRepositoryName = t.TypeOf<typeof LSCRepositoryName>

/**
 * Configuration represents an entire configuration for all
 * LabelSync tools that an organisation is using.
 */
const LSCConfiguration = t.type({
  repos: t.record(LSCRepositoryName, LSCRepository),
})
export interface LSCConfiguration extends t.TypeOf<typeof LSCConfiguration> {}

/**
 * Parses a string into a configuration object.
 * @param input
 */
export function parseConfig(
  input: string,
  logger?: (err: any) => any,
): [string] | [null, LSCConfiguration] {
  try {
    const object = yaml.safeLoad(input)
    const config = LSCConfiguration.decode(object)

    if (config._tag === 'Left') {
      return [PathReporter.report(config).join(os.EOL)]
    }
    return [null, fixConfig(config.right)]
  } catch (err) /* istanbul ignore next */ {
    if (logger) logger(err)
    return [err.message]
  }
}

/**
 * Catches configuration issues that io-ts couldn't catch.
 *
 * @param config
 */
function fixConfig(config: LSCConfiguration): LSCConfiguration {
  const repos = mapEntries(config.repos, fixRepoConfig)

  return {
    ...config,
    repos: repos,
  }
}

/**
 * Fixes issues that io-ts couldn't catch.
 *
 * @param config
 */
function fixRepoConfig(config: LSCRepository): LSCRepository {
  const labels = mapEntries<LSCLabel, LSCLabel>(config.labels, label => ({
    ...label,
    color: fixLabelColor(label.color),
  }))

  return {
    ...config,
    labels,
  }
}

/**
 * Removes hash from the color.
 *
 * @param color
 */
function fixLabelColor(color: string): string {
  if (color.startsWith('#')) return color.slice(1)
  return color
}
