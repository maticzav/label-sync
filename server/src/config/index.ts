import { Plan } from '@prisma/client'

import * as parser from './parser'
import { checkConfiguration } from './checks/configuration'
import { checkPlan } from './checks/plan'
import { getLSConfigRepoName } from './constants'

import * as t from './types'

/**
 * This file bundles up all functions related to configuration.
 */

export * from './constants'
export * from './types'

/**
 * Parses the configuration.
 */
export function parse(params: {
  plan: Plan
  input: string
}): [string] | [null, t.LSCConfiguration] {
  const result = parser.parse(params.input, [
    checkPlan(params.plan),
    checkConfiguration,
  ])

  if (result.success) {
    return [null, result.config]
  }
  return [result.error]
}

// MARK: - Utility functions

/**
 * Returns a list of repositories in configuration without wildcard config.
 */
export function configRepos(config: t.LSCConfiguration): string[] {
  let repos: string[] = []

  for (let repo of config.repos.keys()) {
    repo = repo.trim()
    if (repo !== '*') repos.push(repo)
  }
  return repos
}

/**
 * Tells whether the repository is the configuration
 * repository for the account.
 */
export function isConfigRepo(account: string, repo: string): boolean {
  return getLSConfigRepoName(account) === repo.toLowerCase()
}
