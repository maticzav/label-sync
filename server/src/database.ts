/**
 * This file contains the datasource that connects to the database.
 */

import { PrismaClient } from '@prisma/client'
import * as prisma from '@prisma/client'

import * as t from 'luxon'

import * as c from './config'
import { Maybe } from './data/maybe'

export class Data {
  // MARK: - State

  private prisma: PrismaClient

  // MARK: - Constructor

  constructor(prisma: PrismaClient) {
    this.prisma = prisma
  }

  // MARK: - Methods

  /**
   * Returns the configuration of a particular installation.
   */
  async config(owner: string, repo: string): Promise<Maybe<c.LSCRepository>> {
    const [config] = await this.prisma.repositoryConfiguration.findMany({
      where: {
        configuration: { active: { owner: owner } },
        repository: { in: [repo, '*'] },
      },
      orderBy: {
        repository: 'asc',
      },
      include: {
        labels: true,
      },
    })

    if (config === undefined) return null

    let configuration: c.LSCRepository = {
      config: {
        removeUnconfiguredLabels: config.removeUnconfiguredLabels,
      },
      labels: new Map(),
    }

    for (const label of config.labels) {
      configuration.labels.set(label.name, {
        color: label.color,
        description: label.description,
        alias: label.alias,
        siblings: label.siblings,
      })
    }

    return configuration
  }

  /**
   * Returns the user plan.
   */
  async plan(owner: string): Promise<prisma.Plan> {
    const now = t.DateTime.now()

    const installation = await this.prisma.installation.findUnique({
      where: { owner: owner },
    })

    /**
     * Check that we have an installation and that plan is still valid.
     */
    if (installation === null) return 'FREE'
    if (t.DateTime.fromJSDate(installation.periodEndsAt) < now) return 'FREE'

    return installation.plan
  }

  async install(): Promise<void> {}
}
