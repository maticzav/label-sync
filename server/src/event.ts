/**
 * This file contains a specification for a general event that
 * should be used for adding event handlers to the app.
 *
 * This might seem like a lot of boilerplate; the idea behind separation
 * is that we get a clear instance that works in deterministic way based on
 * the input. You can think of an event as code executing in a very limited
 * environemnt. This file outlines the functions available in it.
 */

import { Probot } from 'probot'
import { ProbotWebhooks } from 'probot/lib/types'

import { Data } from './database'

// MARK: - Spec

// export type Event<E extends WebhookEvents> = (
//   event: E | E[],
//   callback: HandlerFunction<E, U>,
// ) => void

export type Handler = (on: ProbotWebhooks['on'], sources: Sources) => void

/**
 * Sources outline third-party functions available to the event code
 * in the execution environment.
 */
export type Sources = {
  data: IData
}

export interface IData {
  /**
   * Returns the installation associated with a given owner.
   */
  plan: Data['plan']

  /**
   * Returns information about a configuration for a given repository.
   */
  config: Data['config']

  /**
   * Updates the last installation for a given owner.
   */
  install: Data['install']
}

// MARK: - Methods

type BindParams = {
  handlers: Handler[]
  probot: Probot
  sources: Sources
}

/**
 * Binds handlers to sources.
 */
export function bind(params: BindParams) {}
