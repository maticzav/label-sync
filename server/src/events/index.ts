import { Handler } from '../event'

import { handler as installation } from './installation'
import { handler as label } from './label'
import { handler as pullrequest } from './pullrequest'
import { handler as push } from './push'
import { handler as repository } from './repository'

/**
 * Aggregates all the events.
 */
export { handler as installation } from './installation'
export { handler as label } from './label'
export { handler as pullrequest } from './pullrequest'
export { handler as push } from './push'
export { handler as repository } from './repository'

export const handlers: Handler[] = [
  installation,
  label,
  pullrequest,
  push,
  repository,
]
