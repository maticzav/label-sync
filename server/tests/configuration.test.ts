import fs from 'fs'
import path from 'path'

import { parseConfig } from '../src/configuration'

const configurationsPath = path.resolve(
  __dirname,
  './__fixtures__/configurations',
)
const configurations = fs.readdirSync(configurationsPath).map((config) => ({
  config: config,
  path: path.resolve(configurationsPath, config),
}))

describe('configurations:', () => {
  for (const { config, path } of configurations) {
    test(`${config} on FREE`, () => {
      const config = parseConfig(
        {
          id: 'id',
          createdAt: new Date(),
          updatedAt: new Date(),
          email: 'email',
          owner: 'test',
          plan: 'free',
          type: 'USER',
          trial: false,
          planId: 3,
          tier: 'FREE',
        },
        fs.readFileSync(path, { encoding: 'utf-8' }),
      )
      expect(config).toMatchSnapshot()
    })

    test(`${config} on BASIC`, () => {
      const config = parseConfig(
        {
          id: 'id',
          createdAt: new Date(),
          updatedAt: new Date(),
          email: 'email',
          owner: 'test',
          plan: 'free',
          type: 'USER',
          trial: false,
          planId: 3,
          tier: 'BASIC',
        },
        fs.readFileSync(path, { encoding: 'utf-8' }),
      )
      expect(config).toMatchSnapshot()
    })
  }
})
