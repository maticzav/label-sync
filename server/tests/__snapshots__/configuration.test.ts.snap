// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`configurations: anchors.yml on BASIC 1`] = `
Array [
  null,
  Object {
    "colors": Object {
      "area": "#FFD700",
      "bug": "#EE0000",
      "kind": "#3B5BDB",
      "priority": "#F783AC",
      "process": "#EB9100",
      "release": "#A5D8ff",
      "scope": "#27CF79",
      "status": "#F8F9FA",
      "team": null,
    },
    "labels": Object {
      "bug/0-needs-info": Object {
        "color": null,
        "description": "More information is needed for reproduction.",
      },
      "bug/1-repro-available": Object {
        "color": null,
        "description": "A reproduction exists and needs to be confirmed.",
      },
      "bug/2-confirmed": Object {
        "color": null,
        "description": "We have confirmed that this is a bug.",
      },
      "kind/bug": Object {
        "color": null,
        "description": "A reported bug.",
      },
      "kind/discussion": Object {
        "color": null,
        "description": "Discussion is required.",
      },
      "kind/docs": Object {
        "color": null,
        "description": "A documentation change is required.",
      },
      "kind/feature": Object {
        "color": null,
        "description": "A request for a new feature.",
      },
      "kind/improvement": Object {
        "color": null,
        "description": "An improvement to existing feature and code.",
      },
      "kind/question": Object {
        "color": null,
        "description": "Developer asked a question. No code changes required.",
      },
      "kind/regression": Object {
        "color": null,
        "description": "A reported bug in functionality that used to work before.",
      },
      "process/candidate": Object {
        "color": null,
        "description": "Candidate for next Milestone.",
      },
      "process/next-milestone": Object {
        "color": null,
        "description": "Issue earmarked for next Milestone.",
      },
      "process/product": Object {
        "color": null,
        "description": "Temporary label to export products issues from the Engineering process",
      },
    },
    "repos": Object {},
  },
]
`;

exports[`configurations: anchors.yml on FREE 1`] = `
Array [
  null,
  Object {
    "colors": Object {
      "area": "#FFD700",
      "bug": "#EE0000",
      "kind": "#3B5BDB",
      "priority": "#F783AC",
      "process": "#EB9100",
      "release": "#A5D8ff",
      "scope": "#27CF79",
      "status": "#F8F9FA",
      "team": null,
    },
    "labels": Object {
      "bug/0-needs-info": Object {
        "color": null,
        "description": "More information is needed for reproduction.",
      },
      "bug/1-repro-available": Object {
        "color": null,
        "description": "A reproduction exists and needs to be confirmed.",
      },
      "bug/2-confirmed": Object {
        "color": null,
        "description": "We have confirmed that this is a bug.",
      },
      "kind/bug": Object {
        "color": null,
        "description": "A reported bug.",
      },
      "kind/discussion": Object {
        "color": null,
        "description": "Discussion is required.",
      },
      "kind/docs": Object {
        "color": null,
        "description": "A documentation change is required.",
      },
      "kind/feature": Object {
        "color": null,
        "description": "A request for a new feature.",
      },
      "kind/improvement": Object {
        "color": null,
        "description": "An improvement to existing feature and code.",
      },
      "kind/question": Object {
        "color": null,
        "description": "Developer asked a question. No code changes required.",
      },
      "kind/regression": Object {
        "color": null,
        "description": "A reported bug in functionality that used to work before.",
      },
      "process/candidate": Object {
        "color": null,
        "description": "Candidate for next Milestone.",
      },
      "process/next-milestone": Object {
        "color": null,
        "description": "Issue earmarked for next Milestone.",
      },
      "process/product": Object {
        "color": null,
        "description": "Temporary label to export products issues from the Engineering process",
      },
    },
    "repos": Object {},
  },
]
`;

exports[`configurations: basic.yml on BASIC 1`] = `
Array [
  null,
  Object {
    "repos": Object {
      "prisma-test-utils": Object {
        "config": Object {
          "removeUnconfiguredLabels": true,
        },
        "labels": Object {
          "bug/0-needs-reproduction": Object {
            "color": "ff0022",
          },
          "bug/1-has-reproduction": Object {
            "color": "ff0022",
            "description": "Indicates that an issue has reproduction",
          },
          "bug/2-bug-confirmed": Object {
            "color": "red",
          },
          "bug/3-fixing": Object {
            "color": "00ff22",
            "description": "Indicates that we are working on fixing the issue.",
          },
        },
      },
    },
  },
]
`;

exports[`configurations: basic.yml on FREE 1`] = `
Array [
  null,
  Object {
    "repos": Object {
      "prisma-test-utils": Object {
        "config": Object {
          "removeUnconfiguredLabels": true,
        },
        "labels": Object {
          "bug/0-needs-reproduction": Object {
            "color": "ff0022",
          },
          "bug/1-has-reproduction": Object {
            "color": "ff0022",
            "description": "Indicates that an issue has reproduction",
          },
          "bug/2-bug-confirmed": Object {
            "color": "red",
          },
          "bug/3-fixing": Object {
            "color": "00ff22",
            "description": "Indicates that we are working on fixing the issue.",
          },
        },
      },
    },
  },
]
`;

exports[`configurations: duplicate_aliases.yml on BASIC 1`] = `
Array [
  "Configuration references same old label twice in alias:
* \`alias\` appears in repo-two:label/4, label/5

Each legacy label may only be referenced once in alias.",
]
`;

exports[`configurations: duplicate_aliases.yml on FREE 1`] = `
Array [
  "Configuration references same old label twice in alias:
* \`alias\` appears in repo-two:label/4, label/5

Each legacy label may only be referenced once in alias.",
]
`;

exports[`configurations: empty.yml on BASIC 1`] = `
Array [
  null,
  Object {
    "repos": Object {},
  },
]
`;

exports[`configurations: empty.yml on FREE 1`] = `
Array [
  null,
  Object {
    "repos": Object {},
  },
]
`;

exports[`configurations: free.yml on BASIC 1`] = `
Array [
  null,
  Object {
    "repos": Object {
      "emma-cli": Object {
        "labels": Object {
          "bug/0-needs-reproduction": Object {
            "color": "ff0022",
          },
        },
      },
      "label-sync": Object {
        "labels": Object {
          "bug/0-needs-reproduction": Object {
            "color": "ff0022",
          },
        },
      },
      "overthetop": Object {
        "labels": Object {
          "bug/0-needs-reproduction": Object {
            "color": "ff0022",
          },
        },
      },
      "prisma-test-utils": Object {
        "labels": Object {
          "bug/0-needs-reproduction": Object {
            "color": "ff0022",
          },
        },
      },
    },
  },
]
`;

exports[`configurations: free.yml on FREE 1`] = `
Array [
  "You are trying to configure more repositories than there are available in your plan.
Update your current plan to access all the features LabelSync offers.",
]
`;

exports[`configurations: invalid.yml on BASIC 1`] = `
Array [
  "Invalid value undefined supplied to : { repos: { [K in string]: (Partial<{ config: Partial<{ removeUnconfiguredLabels: boolean }> }> & { labels: { [K in string]: ({ color: string } & Partial<{ description: string, siblings: Array<string>, alias: Array<string> }>) } }) } }/repos: { [K in string]: (Partial<{ config: Partial<{ removeUnconfiguredLabels: boolean }> }> & { labels: { [K in string]: ({ color: string } & Partial<{ description: string, siblings: Array<string>, alias: Array<string> }>) } }) }",
]
`;

exports[`configurations: invalid.yml on FREE 1`] = `
Array [
  "Invalid value undefined supplied to : { repos: { [K in string]: (Partial<{ config: Partial<{ removeUnconfiguredLabels: boolean }> }> & { labels: { [K in string]: ({ color: string } & Partial<{ description: string, siblings: Array<string>, alias: Array<string> }>) } }) } }/repos: { [K in string]: (Partial<{ config: Partial<{ removeUnconfiguredLabels: boolean }> }> & { labels: { [K in string]: ({ color: string } & Partial<{ description: string, siblings: Array<string>, alias: Array<string> }>) } }) }",
]
`;

exports[`configurations: missing_siblings.yml on BASIC 1`] = `
Array [
  "Configuration references unconfigured siblings:
* \`label/6\` in repo-one:label/1
* \`label/7\` in repo-one:label/1
* \`label/8\` in repo-two:label/4",
]
`;

exports[`configurations: missing_siblings.yml on FREE 1`] = `
Array [
  "Configuration references unconfigured siblings:
* \`label/6\` in repo-one:label/1
* \`label/7\` in repo-one:label/1
* \`label/8\` in repo-two:label/4",
]
`;

exports[`configurations: multiple_aliases.yml on BASIC 1`] = `
Array [
  "Configuration references multiple old aliases in some label definitions:
* \`repo-one\`:label/1 reference more than one alias

This is the current limitation of LabelSync and will change shortly.",
]
`;

exports[`configurations: multiple_aliases.yml on FREE 1`] = `
Array [
  "Configuration references multiple old aliases in some label definitions:
* \`repo-one\`:label/1 reference more than one alias

This is the current limitation of LabelSync and will change shortly.",
]
`;
